File: implementations/typescript/src/digest.ts
/**
 * The `hif-digest-1` request digest, per specification §14.
 *
 * Used for deduplication, cache keys and stable interaction ids. Never used for
 * matching — two requests with the same digest are the same request, but two
 * requests that *match* need not have the same digest, because matching tolerates
 * variance by design.
 *
 * The expected values in `conformance/cases/digest/` are verified against
 * `openssl dgst -sha256`, not against this code.
 */

import { createHash } from 'node:crypto';
import { canonicalize } from './json/canonical.js';
import { decodeBase64 } from './body.js';
import { encodeQuery, normalizeUrl, decodeQuery } from './url.js';
import { normalizeHeaders } from './headers.js';
import type { HifRequest, JsonValue, QueryParam } from './types.js';

/**
 * Build the value that gets canonicalized and hashed (§14).
 *
 * Exposed separately from `digestRequest` so that a failing conformance case can
 * be debugged by looking at the pre-image rather than guessing at a hash.
 */
export function digestPreimage(req: HifRequest): JsonValue {
  const url = normalizeUrl(req.url);

  // §14: query parameters sorted by (name, value) using UTF-16 code unit order.
  const params = [...decodeQuery(url.rawQuery)].sort(compareParams);
  const authority = url.port === null ? url.host : `${url.host}:${url.port}`;
  const query = encodeQuery(params);
  const u = `${url.scheme}://${authority}${url.path}${query === '' ? '' : `?${query}`}`;

  // §14: headers lowercased and sorted by (name, value); repeats produce repeats.
  const headers = normalizeHeaders(req.headers)
    .map((h): [string, string] => [h.name, h.value])
    .sort((a, b) => compareUtf16(a[0], b[0]) || compareUtf16(a[1], b[1]));

  const body = req.body;
  let b: JsonValue;
  if (!body || body.encoding === 'empty') b = null;
  else if (body.encoding === 'text') b = body.text;
  else if (body.encoding === 'json') b = body.json;
  else b = { base64: body.base64 };

  // Single-letter member names so that RFC 8785's name sort produces b, h, m, u
  // in that order, which makes the pre-image readable by hand.
  return { b, h: headers as unknown as JsonValue, m: req.method, u };
}

/** The lowercase hex SHA-256 of the canonical pre-image. */
export function digestRequest(req: HifRequest): string {
  const preimage = canonicalize(digestPreimage(req));
  return createHash('sha256').update(preimage, 'utf8').digest('hex');
}

/**
 * Validate that a base64 body decodes, and return its byte length.
 * Used by `spool inspect` to report sizes without materialising the bytes twice.
 */
export function base64ByteLength(b64: string): number {
  return decodeBase64(b64).length;
}

function compareParams(a: QueryParam, b: QueryParam): number {
  return compareUtf16(a.name, b.name) || compareUtf16(a.value, b.value);
}

/**
 * UTF-16 code unit ordering.
 *
 * JavaScript's `<` on strings is already UTF-16 code unit ordering, which is
 * what RFC 8785 and §14 specify. `localeCompare` is emphatically not, and using
 * it here would make digests locale-dependent.
 */
function compareUtf16(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

File: implementations/python/src/spool/digest.py
"""The ``hif-digest-1`` request digest, per specification section 14.

Used for deduplication, cache keys and stable interaction ids. Never used for
matching: two requests with the same digest are the same request, but two
requests that *match* need not have the same digest, because matching tolerates
variance by design.

The expected values in ``conformance/cases/digest/`` are verified against
``openssl dgst -sha256``, not against this code.
"""

from __future__ import annotations

import hashlib
from typing import Any, Dict, List

from .body import normalize_headers
from .canonical import canonicalize
from .url import QueryParam, decode_query, encode_query, normalize_url


def digest_preimage(request: Dict[str, Any]) -> Dict[str, Any]:
    """Build the value that gets canonicalized and hashed (section 14).

    Exposed separately from :func:`digest_request` so a failing conformance case
    can be debugged by looking at the pre-image rather than guessing at a hash.
    """
    url = normalize_url(request["url"])

    # Section 14: query parameters sorted by (name, value), UTF-16 code unit order.
    params: List[QueryParam] = sorted(decode_query(url.raw_query), key=_param_sort_key)
    netloc = url.host if url.port is None else f"{url.host}:{url.port}"
    query = encode_query(params)
    normalized_url = f"{url.scheme}://{netloc}{url.path}" + (f"?{query}" if query else "")

    # Section 14: headers lowercased and sorted by (name, value); repeats repeat.
    headers = sorted(
        ([h.name, h.value] for h in normalize_headers(request.get("headers"))),
        key=lambda pair: (_utf16_key(pair[0]), _utf16_key(pair[1])),
    )

    body = request.get("body")
    encoding = (body or {}).get("encoding")
    if not body or encoding == "empty":
        payload: Any = None
    elif encoding == "text":
        payload = body["text"]
    elif encoding == "json":
        payload = body["json"]
    else:
        payload = {"base64": body["base64"]}

    # Single-letter member names so RFC 8785's name sort produces b, h, m, u in
    # that order, which makes the pre-image readable by hand.
    return {"b": payload, "h": headers, "m": request["method"], "u": normalized_url}


def digest_request(request: Dict[str, Any]) -> str:
    """The lowercase hex SHA-256 of the canonical pre-image."""
    preimage = canonicalize(digest_preimage(request))
    return hashlib.sha256(preimage.encode("utf-8")).hexdigest()


def _param_sort_key(param: QueryParam) -> Any:
    return (_utf16_key(param.name), _utf16_key(param.value))


def _utf16_key(text: str) -> Any:
    """UTF-16 code unit ordering, matching RFC 8785 and JavaScript string sort."""
    encoded = text.encode("utf-16-be", errors="surrogatepass")
    return tuple(int.from_bytes(encoded[i : i + 2], "big") for i in range(0, len(encoded), 2))
